components/server/ee/src/workspace/gitpod-server-impl.ts

```diff
@@ -1304,9 +1377,107 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
+        const now = new Date();
+        const teamSubscription = await this.teamSubscription2DB.findForTeam(teamId, now.toISOString());
+        if (!teamSubscription) {
+            // No team subscription, nothing to do 🌴
+            return;
+        }
+        await this.updateTeamSubscriptionQuantity(teamSubscription);
+        await this.teamSubscription2Service.cancelTeamMemberSubscription(
+            teamSubscription,
+            userId,
+            teamMembershipId,
+            now,
+        );
+    }
+
+    protected async onTeamDeleted(teamId: string): Promise<void> {
+        const now = new Date();
+        const teamSubscription = await this.teamSubscription2DB.findForTeam(teamId, now.toISOString());
+        if (!teamSubscription) {
+            // No team subscription, nothing to do 🌴
+            return;
+        }
+        const chargebeeSubscriptionId = teamSubscription.paymentReference;
+        await this.chargebeeService.cancelSubscription(
+            chargebeeSubscriptionId,
+            {},
+            { teamId, chargebeeSubscriptionId },
+        );
+    }
+
+    protected async updateTeamSubscriptionQuantity(teamSubscription: TeamSubscription2): Promise<void> {
+        const members = await this.teamDB.findMembersByTeam(teamSubscription.teamId);
+        const oldQuantity = teamSubscription.quantity;
+        const newQuantity = members.length;
+        log.info(`updateTeamSubscriptionQuantity: oldQuantity=${oldQuantity}, newQuantity=${newQuantity}`);
+        try {
+            if (oldQuantity < newQuantity) {
+                // Upgrade: Charge for it!
+                const chargebeeSubscription = await this.getChargebeeSubscription(
+                    {},
+                    teamSubscription.paymentReference,
+                );
+                let pricePerUnitInCents = chargebeeSubscription.plan_unit_price;
+                if (pricePerUnitInCents === undefined) {
+                    const plan = Plans.getById(teamSubscription.planId)!;
+                    pricePerUnitInCents = plan.pricePerMonth * 100;
+                }
+                const currentTermRemainingRatio =
+                    this.upgradeHelper.getCurrentTermRemainingRatio(chargebeeSubscription);
+                const diffInCents = Math.round(
+                    pricePerUnitInCents * (newQuantity - oldQuantity) * currentTermRemainingRatio,
+                );
+                const upgradeTimestamp = new Date().toISOString();
+                const description = `Pro-rated upgrade from '${oldQuantity}' to '${newQuantity}' team members (${formatDate(
+                    upgradeTimestamp,
+                )})`;
+                log.info(
+                    `chargeForUpgrade: paymentReference=${teamSubscription.paymentReference}, currentTermRemainingRatio=${currentTermRemainingRatio}, diffInCents=${diffInCents}`,
+                );
+                await this.upgradeHelper.chargeForUpgrade(
+                    "",
```

Comment:
why `""`?

Reply:
@AlexTugarev Would you like this to be changed/improved somehow?